File: 2176-count-equal-and-divisible-pairs-in-an-array/2176-count-equal-and-divisible-pairs-in-an-array.py
class Solution:
    def countPairs(self, nums: List[int], k: int) -> int:
        # Agorithm: Brute Force

        # Input: array nums and an integer k
        # Output: Count
        # Coditions:
        # nums[i] == nums[j]
        # (i * j) is divisible by k

        count = 0
        n = len(nums)
        for i in range(n):
            for j in range(i + 1, n):
                # Just copy past explanation part
                if nums[i] != nums[j]: continue
                if (i * j) % k == 0:
                    count += 1
        return count
        

File: 2176-count-equal-and-divisible-pairs-in-an-array/2176-count-equal-and-divisible-pairs-in-an-array.java
class Solution {
    public int countPairs(int[] nums, int k) {
        int count = 0;
        Map<Integer, List<Integer>> indexMap = new HashMap<>();
        for (int j = 0; j < nums.length; j++) {
            int num = nums[j];
            if (indexMap.containsKey(num)) {
                for (int i : indexMap.get(num)) {
                    if ((long)i * j % k == 0) {
                        count++; // += 1
                    }
                }
            }
            indexMap.computeIfAbsent(num, key -> new ArrayList<>()).add(j);
        }
        return count;
    }
}